File: bmi/bmiheat_ex.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "bmi_interoperability.h"


int main(int argc, char *argv[]) {
  int model, status, n_invars, n_outvars, i, j, k;
  char *config_file = "test.cfg";
  int nchars = strlen(config_file);
  char *component_name;
  char **input_var_names, **output_var_names;
  double time;
  char *units;
  char *var_name;
  int grid_id;
  char *grid_type;
  int grid_rank;
  int *grid_shape;
  int grid_size;
  char *var_type;
  int var_nbytes;
  void *buffer;
  void *new;
  void *ref;

  // Get a new model.
  model = new_model();
  printf("Model %d\n", model);

  // Initialize the model.
  status = initialize(model, config_file, nchars);
  printf("- initialize status: %d\n", status);

  // Get the model's name.
  component_name = malloc(MAX_COMPONENT_NAME);
  status = get_component_name(model, component_name, MAX_COMPONENT_NAME);
  printf("- component name: %s\n", component_name);
  free(component_name);

  // Get the model's inputs. Pointers to pointers, oh my.
  status = get_input_var_name_count(model, &n_invars);
  printf("- number of input vars: %d\n", n_invars);
  input_var_names = malloc(n_invars);
  input_var_names[0] = malloc(MAX_VAR_NAME);
  for (i = 1; i < n_invars; i++) {
    input_var_names[i] = input_var_names[i - 1] + MAX_VAR_NAME;
  }
  status = get_input_var_names(model, input_var_names, n_invars);
  printf("- input var names:\n");
  for (i = 0; i < n_invars; i++) {
    printf("  - %s\n", input_var_names[i]);
  }
  free(input_var_names);

  // Get the model's outputs.
  status = get_output_var_name_count(model, &n_outvars);
  printf("- number of output vars: %d\n", n_outvars);
  output_var_names = malloc(n_outvars);
  output_var_names[0] = malloc(MAX_VAR_NAME);
  for (i = 1; i < n_outvars; i++) {
    output_var_names[i] = output_var_names[i - 1] + MAX_VAR_NAME;
  }
  status = get_output_var_names(model, output_var_names, n_outvars);
  printf("- output var names:\n");
  for (i = 0; i < n_outvars; i++) {
    printf("  - %s\n", output_var_names[i]);
  }
  free(output_var_names);

  // Get time information from the model.
  status = get_start_time(model, &time);
  printf("- model start time: %6.1f\n", time);
  status = get_end_time(model, &time);
  printf("- model stop time: %6.1f\n", time);
  status = get_current_time(model, &time);
  printf("- model current time: %6.1f\n", time);
  status = get_time_step(model, &time);
  printf("- model time step: %6.1f\n", time);
  units = malloc(MAX_UNITS_NAME);
  status = get_time_units(model, units, MAX_UNITS_NAME);
  printf("- model time units: %s\n", units);
  free(units);

  // Update the model by a single time step.
  status = update(model);
  printf("- update status: %d\n", status);
  status = get_current_time(model, &time);
  printf("- model current time: %6.1f\n", time);

  // Update the model by a fraction of a time step.
  status = update_frac(model, 0.75);
  printf("- update_frac status: %d\n", status);
  status = get_current_time(model, &time);
  printf("- model current time: %6.1f\n", time);

  // Update the model until a later time.
  status = update_until(model, 10.0);
  printf("- update_until status: %d\n", status);
  status = get_current_time(model, &time);
  printf("- model current time: %6.1f\n", time);

  // Get grid information for the plate_surface__temperature variable.
  var_name = "plate_surface__temperature";
  nchars = strlen(var_name);
  status = get_var_grid(model, var_name, nchars, &grid_id);
  printf("- Grid info for %s\n", var_name);
  printf(" - grid_id: %d\n", grid_id);
  grid_type = malloc(MAX_TYPE_NAME);
  memset(grid_type, 0, MAX_TYPE_NAME);
  status = get_grid_type(model, grid_id, grid_type, MAX_TYPE_NAME);
  printf(" - grid type: %s\n", grid_type);
  free(grid_type);
  status = get_grid_rank(model, grid_id, &grid_rank);
  printf(" - grid rank: %d\n", grid_rank);
  grid_shape = malloc(grid_rank * sizeof(int));
  /* status = get_grid_shape(model, grid_id, &grid_shape[0]); */
  status = get_grid_shape(model, grid_id, grid_shape, grid_rank);
  printf(" - grid shape:");
  for (i = 0; i < grid_rank; i++) {
    printf(" %d", grid_shape[i]);
  }
  printf("\n");
  status = get_grid_size(model, grid_id, &grid_size);
  printf(" - grid size: %d\n", grid_size);

  // Get information for the plate_surface__temperature variable.
  var_name = "plate_surface__temperature";
  nchars = strlen(var_name);
  printf("- Variable info for %s\n", var_name);
  var_type = malloc(MAX_TYPE_NAME);
  memset(var_type, 0, MAX_TYPE_NAME);
  status = get_var_type(model, var_name, nchars, var_type, MAX_TYPE_NAME);
  printf(" - variable type: %s\n", var_type);
  free(var_type);
  status = get_var_nbytes(model, var_name, nchars, &var_nbytes);
  printf(" - total memory (bytes): %d\n", var_nbytes);

  // Get the values. (Note: not completely confident gridded order is correct.)
  buffer = malloc(grid_size * sizeof(float));
  status = get_value_float(model, var_name, nchars, buffer, grid_size);
  printf(" - values, streamwise:\n");
  for (i = 0; i < grid_size; i++) {
    printf(" %6.1f", ((float *)buffer)[i]);
  }
  printf(" - values, dimensional:\n");
  for (j = 0; j < grid_shape[1]; j++) {
    for (i = 0; i < grid_shape[0]; i++) {
      k = i + j*grid_shape[0];
      printf(" %6.1f", ((float *)buffer)[k]);
    }
    printf("\n");
  }
  free(buffer);

  // Set values (float).
  new = malloc(grid_size * sizeof(float));
  for (i = 0; i < grid_size; i++) {
    ((float *)new)[i] = (float)i;
  }
  status = set_value_float(model, var_name, nchars, new, grid_size);
  free(new);
  buffer = malloc(grid_size * sizeof(float));
  status = get_value_float(model, var_name, nchars, buffer, grid_size);
  printf(" - new values (set/get), streamwise:\n");
  for (i = 0; i < grid_size; i++) {
    printf(" %6.1f", ((float *)buffer)[i]);
  }
  free(buffer);

  // Get a reference to a (float) variable and check that it updates.
  status = get_current_time(model, &time);
  status = get_value_ptr(model, var_name, nchars, &ref);
  printf(" - values, dimensional, through ref, at time %6.1f:\n", time);
  for (j = 0; j < grid_shape[1]; j++) {
    for (i = 0; i < grid_shape[0]; i++) {
      k = i + j*grid_shape[0];
      printf(" %6.1f", ((float *)ref)[k]);
    }
    printf("\n");
  }
  status = update(model);
  printf("- update status: %d\n", status);
  status = get_current_time(model, &time);
  printf(" - values, dimensional, through ref, at time %6.1f:\n", time);
  for (j = 0; j < grid_shape[1]; j++) {
    for (i = 0; i < grid_shape[0]; i++) {
      k = i + j*grid_shape[0];
      printf(" %6.1f", ((float *)ref)[k]);
    }
    printf("\n");
  }

  // Clean up variables.
  free(grid_shape);

  // Finalize the model.
  status = finalize(model);
  printf("- finalize status: %d\n", status);
}

File: bmi/bmiheat_ex.py
"""An example of calling a Fortran BMI through Cython."""

import numpy as np
from bmi_heat import Heat


config_file = 'test.cfg'


np.set_printoptions(formatter={'float': '{: 6.1f}'.format})

# Instantiate a model and get its name.
m = Heat()
print(m.get_component_name())

# Initialize the model.
m.initialize(config_file)

# List the model's echange items.
print('Input vars:', m.get_input_var_names())
print('Output vars:', m.get_output_var_names())

# Get time information from the model.
print('Start time:', m.get_start_time())
print('End time:', m.get_end_time())
print('Current time:', m.get_current_time())
print('Time step:', m.get_time_step())
print('Time units:', m.get_time_units())

# Advance the model by one time step.
m.update()
print('Current time:', m.get_current_time())

# Advance the model by a fractional time step.
m.update_frac(0.75)
print('Current time:', m.get_current_time())

# Advance the model until a later time.
m.update_until(10.0)
print('Current time:', m.get_current_time())

# Get the grid_id for the plate_surface__temperature variable.
var_name = 'plate_surface__temperature'
print('Variable {}'.format(var_name))
grid_id = m.get_var_grid(var_name)
print(' - grid id:', grid_id)

# Get grid and variable info for plate_surface__temperature.
print(' - grid type:', m.get_grid_type(grid_id))
grid_rank = m.get_grid_rank(grid_id)
print(' - rank:', grid_rank)
grid_shape = np.empty(grid_rank, dtype=np.int32)
m.get_grid_shape(grid_id, grid_shape)
print(' - shape:', grid_shape)
grid_size = m.get_grid_size(grid_id)
print(' - size:', grid_size)
grid_spacing = np.empty(grid_rank, dtype=np.float64)
m.get_grid_spacing(grid_id, grid_spacing)
print(' - spacing:', grid_spacing)
grid_origin = np.empty(grid_rank, dtype=np.float64)
m.get_grid_origin(grid_id, grid_origin)
print(' - origin:', grid_origin)
print(' - variable type:', m.get_var_type(var_name))
print(' - units:', m.get_var_units(var_name))
print(' - itemsize:', m.get_var_itemsize(var_name))
print(' - nbytes:', m.get_var_nbytes(var_name))

# Get the temperature values.
val = np.empty(grid_shape, dtype=np.float32)
m.get_value(var_name, val)
print(' - values (streamwise):')
print(val)
print(' - values (gridded):')
print(val.reshape(np.roll(grid_shape, 1)))

# Set new temperature values.
new = np.arange(grid_size, dtype=np.float32)  # 'real*4 in Fortran
m.set_value(var_name, new)
check = np.empty(grid_shape, dtype=np.float32)
m.get_value(var_name, check)
print(' - new values (set/get, streamwise):');
print(check)

# Get a reference to the temperature values and check that it updates.
print(' - values (by ref, streamwise) at time {}:'.format(m.get_current_time()))
ref = m.get_value_ptr(var_name)
print(ref)
m.update()
print(' - values (by ref, streamwise) at time {}:'.format(m.get_current_time()))
print(ref)

# Get the grid_id for the plate_surface__thermal_diffusivity variable.
var_name = 'plate_surface__thermal_diffusivity'
print('Variable {}'.format(var_name))
grid_id = m.get_var_grid(var_name)
print(' - grid id:', grid_id)

# Get grid and variable info for plate_surface__thermal_diffusivity.
print(' - grid type:', m.get_grid_type(grid_id))
grid_rank = m.get_grid_rank(grid_id)
print(' - rank:', grid_rank)
print(' - size:', m.get_grid_size(grid_id))
grid_x = np.empty(max(grid_rank, 1), dtype=np.float64)
m.get_grid_x(grid_id, grid_x)
print(' - x:', grid_x)
grid_y = np.empty(max(grid_rank, 1), dtype=np.float64)
m.get_grid_y(grid_id, grid_y)
print(' - y:', grid_y)
grid_z = np.empty(max(grid_rank, 1), dtype=np.float64)
m.get_grid_z(grid_id, grid_z)
print(' - z:', grid_z)
grid_connectivity = np.empty(max(grid_rank, 1), dtype=np.int32)
m.get_grid_connectivity(grid_id, grid_connectivity)
print(' - connectivity:', grid_connectivity)
grid_offset = np.empty(max(grid_rank, 1), dtype=np.int32)
m.get_grid_offset(grid_id, grid_offset)
print(' - offset:', grid_offset)
print(' - variable type:', m.get_var_type(var_name))
print(' - units:', m.get_var_units(var_name))
print(' - itemsize:', m.get_var_itemsize(var_name))
print(' - nbytes:', m.get_var_nbytes(var_name))

# Get the diffusivity values.
val = np.empty(1, dtype=np.float32)
m.get_value(var_name, val)
print(' - values:')
print(val)

# Get the grid_id for the model__identification_number variable.
var_name = 'model__identification_number'
print('Variable {}'.format(var_name))
grid_id = m.get_var_grid(var_name)
print(' - grid id:', grid_id)

# Get grid and variable info for model__identification_number.
print(' - grid type:', m.get_grid_type(grid_id))
grid_rank = m.get_grid_rank(grid_id)
print(' - rank:', grid_rank)
print(' - size:', m.get_grid_size(grid_id))
grid_x = np.empty(max(grid_rank, 1), dtype=np.float64)
m.get_grid_x(grid_id, grid_x)
print(' - x:', grid_x)
grid_y = np.empty(max(grid_rank, 1), dtype=np.float64)
m.get_grid_y(grid_id, grid_y)
print(' - y:', grid_y)
grid_z = np.empty(max(grid_rank, 1), dtype=np.float64)
m.get_grid_z(grid_id, grid_z)
print(' - z:', grid_z)
grid_connectivity = np.empty(max(grid_rank, 1), dtype=np.int32)
m.get_grid_connectivity(grid_id, grid_connectivity)
print(' - connectivity:', grid_connectivity)
grid_offset = np.empty(max(grid_rank, 1), dtype=np.int32)
m.get_grid_offset(grid_id, grid_offset)
print(' - offset:', grid_offset)
print(' - variable type:', m.get_var_type(var_name))
print(' - units:', m.get_var_units(var_name))
print(' - itemsize:', m.get_var_itemsize(var_name))
print(' - nbytes:', m.get_var_nbytes(var_name))

# Get the model id.
val = np.empty(1, dtype=np.int32)
m.get_value(var_name, val)
print(' - values:')
print(val)

# Set new model id.
new = np.array(42, dtype=np.intc)
m.set_value(var_name, new)
check = np.empty(1, dtype=np.int32)
m.get_value(var_name, check)
print(' - new values (set/get):');
print(check)

# Finalize the model.
m.finalize()

# Check that number of instances can't exceed N_MODELS=3.
# a = Heat()
# b = Heat()
# c = Heat()  # should fail with index=-1
